python/ray/experimental/data/dataset.py

```diff
@@ -195,6 +196,117 @@ def repartition(self, num_blocks: int) -> "Dataset[T]":
         new_blocks = simple_shuffle(self._blocks, num_blocks)
         return Dataset(new_blocks)
 
+    def split(self, n: int,
+              locality_hints: List[Any] = None) -> List["Dataset[T]"]:
+        """Split the dataset into ``n`` disjoint pieces.
+
+        This returns a list of sub-datasets that can be passed to Ray tasks
+        and actors and used to read the dataset records in parallel.
+
+        Examples:
+            >>> # Split up a dataset to process over `n` worker actors.
+            >>> shards = ds.split(len(workers), locality_hints=workers)
+            >>> for shard, worker in zip(shards, workers):
+            ...     worker.consume.remote(shard)
+
+        Time complexity: O(1)
+
+        Args:
+            n: Number of child datasets to return.
+            locality_hints: A list of Ray actor handles of size ``n``. The
+                system will try to co-locate the blocks of the ith dataset
+                with the ith actor to maximize data locality.
+
+        Returns:
+            A list of ``n`` disjoint dataset splits.
+        """
+        assert n > 0 and n <= len(self._blocks)
+        assert locality_hints is None or len(locality_hints) == n
+
+        # Split self._blocks into n chunks. the first {num_large_chunks} chunks
+        # have {small_chunk_size + 1} blocks, and remaining
+        # chunks have {small_chunk_size} blocks.
+        small_chunk_size = len(self._blocks) // n
+        num_large_chunks = len(self._blocks) - small_chunk_size * n
+
+        datasets = []
+        block_refs = list(self._blocks)
+
+        if locality_hints is None:
+            cur = 0
+            for i in range(n):
+                end = cur + small_chunk_size
+                if i < num_large_chunks:
+                    end = end + 1
+                datasets.append(Dataset(block_refs[cur:end]))
+                cur = end
+            return datasets
+
+        # If the locality_hints is set, we use a greedy algorithm
+        # to co-locate the blocks with the actors based on block
+        # and actor's node_id:
+        #
+        # 1. We get the node_id where the block resides in and store it in
+        # a map {node_id: [blocks]}. Note that each block could
+        # be located on multiple nodes. For simplicity, we use the first
+        # node_id so that there is one to one mapping from node_id to block
+        # in the map.
+        #
+        # 2. We do two round of allocations. In the first round, we allocate
+        # blocks for each actor based by sustract the blocks from the map
+        # we build from previous step.
+        #
+        # 3. Some actors might have more blocks than needed while the
+        # others don't have enough blocks. We do second round of allocation
+        # for the reminder blocks so that every actor have sufficient blocks.
+
+        # wait might affect the throughput due to stragglers?
+        ray.wait(block_refs)
```

Comment:
per @rkooo567 's comments. we need to decide if we want to wait for all blocks or not wait at all?

Reply:
How about we wait for all the blocks if `locality_hints` is given, otherwise don't do anything?

Reply:
+1 on this. We have this just because get_object_locations require objects to be resolved right?